server/src/handlers/sql.rs

```diff
@@ -128,26 +145,100 @@ fn convert_output(output: Output) -> ArrowResult<Response> {
 }
 
 fn convert_records(records: RecordBatchVec) -> ArrowResult<Response> {
-    let total_rows = records.iter().map(|v| v.num_rows()).sum();
-    let mut resp = Vec::with_capacity(total_rows);
+    // let total_rows = records.iter().map(|v| v.num_rows()).sum();
+    // let mut resp = Vec::with_capacity(total_rows);
+
+    let mut response_rows = Vec::new();
     for record_batch in records {
         let num_cols = record_batch.num_columns();
         let num_rows = record_batch.num_rows();
         let schema = record_batch.schema();
 
+        let mut column_names = Vec::with_capacity(num_cols);
+        let mut column_data = Vec::with_capacity(num_rows);
+
+        for col_idx in 0..num_cols {
+            let column_schema = schema.column(col_idx).clone();
+            column_names.push(ResponseColumn {
+                name: column_schema.name,
+                data_type: column_schema.data_type,
+            });
+        }
+
         for row_idx in 0..num_rows {
-            let mut row = HashMap::with_capacity(num_cols);
+            let mut row_data = Vec::with_capacity(num_cols);
             for col_idx in 0..num_cols {
                 let column = record_batch.column(col_idx);
                 let column = column.datum(row_idx);
 
-                let column_name = schema.column(col_idx).name.clone();
-                row.insert(column_name, column);
+                row_data.push(column);
             }
 
-            resp.push(row);
+            column_data.push(row_data);
+        }
+
+        response_rows.push(ResponseRow {
+            column_names,
+            data: column_data,
+        })
+    }
+
+    Ok(Response::Rows(response_rows))
+}
+
+mod http_format {
+    use serde::{
+        ser::SerializeSeq,
+        Serializer,
+    };
+
+    use super::Row;
+    use crate::handlers::sql::ResponseRows;
+
+    pub fn serialize<S>(records: &ResponseRows, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let total_count = records.iter().map(|row| row.data.len()).sum();
+        let mut seq = serializer.serialize_seq(Some(total_count))?;
+
+        for record in records {
+            let column_names = &record.column_names;
+            // let mut map = seq.serialize_map(Some(column_names.len()))?;
+
+            for rows in &record.data {
+                let data = rows
+                    .iter()
+                    .enumerate()
+                    .map(|(col_idx, datum)| {
+                        let column_name = &column_names[col_idx].name.clone();
+                        (column_name.clone(), datum.clone())
```

Comment:
Those clones might be avoidable. Maybe with some lifetime annotations on `Row` 🧐